apps/lottie-preview/src/configUtils.ts

```diff
@@ -19,59 +29,197 @@ export interface JsonField {
+  try {
+    return await fn();
+  } catch (error: any) {
+    if (retries > 0 && (error.message?.includes('rate limit') || error.message?.includes('429'))) {
+      console.warn(`Rate limit hit, retrying in ${delay}ms... (${retries} retries left)`);
+      await sleep(delay);
+      return retryWithBackoff(fn, retries - 1, delay * 2); // Exponential backoff
     }
+    throw error;
+  }
+}
+
+// Add timeout wrapper for the entire operation
+const withTimeout = <T>(promise: Promise<T>, timeoutMs: number): Promise<T> => {
+  return Promise.race([promise, new Promise<T>((_, reject) => setTimeout(() => reject(new Error(`Operation timed out after ${timeoutMs}ms`)), timeoutMs))]);
+};
+
+export async function getContentTypesWithJsonFieldsCount(cma: ConfigAppSDK['cma']): Promise<number> {
+  try {
+    // Fetch ALL content types (lightweight, no editor interfaces)
+    const contentTypes = await cma.contentType.getMany({
+      query: { limit: 1000 },
+    });
+
+    let count = 0;
+    for (const contentType of contentTypes.items) {
+      const hasJsonFields = contentType.fields.some((f) => f.type === JsonFieldType);
+      if (hasJsonFields) {
+        count++;
+      }
+    }
+
+    return count;
+  } catch (error: any) {
+    console.error('getContentTypesWithJsonFieldsCount failed:', error.message);
+    return 0;
   }
+}
+
+export async function getJsonFields(
+  cma: ConfigAppSDK['cma'],
+  appDefinitionId: string,
+  options?: { limit?: number; offset?: number },
+  onProgress?: (processed: number, total: number) => void
+): Promise<JsonFieldsResult> {
+  try {
+    const result = await withTimeout(getJsonFieldsInternal(cma, appDefinitionId, options, onProgress), OVERALL_TIMEOUT);
+    return result;
+  } catch (error: any) {
+    console.error('getJsonFields failed:', error.message);
+    return {
+      fields: [],
+      totalContentTypes: 0,
+      processedContentTypes: 0,
+      hasMore: false,
+    };
+  }
+}
+
+async function getJsonFieldsInternal(
+  cma: ConfigAppSDK['cma'],
+  appDefinitionId: string,
+  options?: { limit?: number; offset?: number },
+  onProgress?: (processed: number, total: number) => void
+): Promise<JsonFieldsResult> {
+  const contentTypes = await cma.contentType.getMany({
+    query: { limit: 1000 },
+  });
+
+  // Filter to only content types that have JSON Object fields
+  const contentTypesWithJson = contentTypes.items
+    .map((contentType) => ({
+      contentType,
+      jsonFields: contentType.fields.filter((f) => f.type === JsonFieldType),
+    }))
+    .filter(({ jsonFields }) => jsonFields.length > 0);
+
+  if (contentTypesWithJson.length === 0) {
+    return {
+      fields: [],
+      totalContentTypes: contentTypes.items.length,
+      processedContentTypes: 0,
+      hasMore: false,
+    };
+  }
+
+  const offset = options?.offset || 0;
+  const limit = options?.limit || INITIAL_BATCH_SIZE;
+  const totalWithJsonFields = contentTypesWithJson.length;
+
+  const startIndex = offset;
```

Comment:
[nit pick] Could use `offset` instead of creating this alias variable `startIndex` - or rename `offset` to `startIndex` if it's better for readability